batch/sql/add-billing-limit.sql

```diff
@@ -0,0 +1,129 @@
+  JOIN (SELECT billing_project, resource, quantity
+        FROM attempt_resources
+        LEFT JOIN batches ON batches.id = attempt_resources.batch_id
+        WHERE batch_id = NEW.batch_id AND
+        job_id = NEW.job_id AND
+        attempt_id = NEW.attempt_id) AS t
+  ON aggregated_billing_project_resources.billing_project = t.billing_project AND aggregated_billing_project_resources.resource = t.resource
+  SET `usage` = `usage` + msec_diff * t.quantity
+  WHERE aggregated_billing_project_resources.billing_project = cur_billing_project;
+
+  UPDATE aggregated_batch_resources
+  JOIN (SELECT batch_id, resource, quantity
+        FROM attempt_resources
+        WHERE batch_id = NEW.batch_id AND
+        job_id = NEW.job_id AND
+        attempt_id = NEW.attempt_id) AS t
+  ON aggregated_batch_resources.batch_id = t.batch_id AND
+    aggregated_batch_resources.resource = t.resource
+  SET `usage` = `usage` + msec_diff * t.quantity
+  WHERE aggregated_batch_resources.batch_id = NEW.batch_id;
+
+  UPDATE aggregated_job_resources
+  JOIN (SELECT batch_id, job_id, resource, quantity
+        FROM attempt_resources
+        WHERE batch_id = NEW.batch_id AND
+        job_id = NEW.job_id AND
+        attempt_id = NEW.attempt_id) AS t
+  ON aggregated_job_resources.batch_id = t.batch_id AND
+    aggregated_job_resources.job_id = t.job_id AND
+    aggregated_job_resources.resource = t.resource
+  SET `usage` = `usage` + msec_diff * t.quantity
+  WHERE aggregated_job_resources.batch_id = NEW.batch_id AND aggregated_job_resources.job_id = NEW.job_id;
+END $$
+
+DROP TRIGGER IF EXISTS attempt_resources_after_insert $$
+CREATE TRIGGER attempt_resources_after_insert AFTER INSERT ON attempt_resources
+FOR EACH ROW
+BEGIN
+  DECLARE cur_start_time BIGINT;
+  DECLARE cur_end_time BIGINT;
+  DECLARE cur_billing_project VARCHAR(100);
+  DECLARE msec_diff BIGINT;
+
+  SELECT billing_project INTO cur_billing_project FROM batches WHERE id = NEW.batch_id;
+
+  SELECT start_time, end_time INTO cur_start_time, cur_end_time
+  FROM attempts
+  WHERE batch_id = NEW.batch_id AND job_id = NEW.job_id AND attempt_id = NEW.attempt_id
+  LOCK IN SHARE MODE;
+
+  SET msec_diff = GREATEST(COALESCE(cur_end_time - cur_start_time, 0), 0);
+
+  INSERT INTO aggregated_job_resources (batch_id, job_id, resource, `usage`)
+  VALUES (NEW.batch_id, NEW.job_id, NEW.resource, NEW.quantity * msec_diff)
+  ON DUPLICATE KEY UPDATE
+    `usage` = `usage` + NEW.quantity * msec_diff;
+
+  INSERT INTO aggregated_batch_resources (batch_id, resource, `usage`)
+  VALUES (NEW.batch_id, NEW.resource, NEW.quantity * msec_diff)
+  ON DUPLICATE KEY UPDATE
+    `usage` = `usage` + NEW.quantity * msec_diff;
+
+  INSERT INTO aggregated_billing_project_resources (billing_project, resource, `usage`)
+  VALUES (cur_billing_project, NEW.resource, NEW.quantity * msec_diff)
+  ON DUPLICATE KEY UPDATE
+    `usage` = `usage` + NEW.quantity * msec_diff;
+END $$
+
+DELIMITER ;
+
+UPDATE billing_projects
```

Comment:
My one concern is how to block any updates from occurring while these two queries run.